functions/calcTajimaD.txt

```diff
@@ -0,0 +1,65 @@
+// Written by Nick Bailey, submitted on 2024-06-13
+// Currently affiliated with CNRS and University of Lyon 1
+// This code is provided in the public domain free of restriction, please credit SLiM-Extras (https://github.com/MesserLab/SLiM-Extras) and myself, Nick Bailey, if you use it
+
+// calculate neutrality test based on the site frequency spectrum, Tajima's D
+// calculations originally in Tajima 1989
+function (float)calcTajimaD(object<Genome> genomes, [No<Mutation> muts = NULL], [Ni$ start = NULL], [Ni$ end = NULL])
+// conduct various checks for proper usage copied straight from calcWattersonTheta()
+{ if (genomes.length() == 0)
+		stop("ERROR (calcTajimaD()): genomes must be non-empty.");
+	if (community.allSpecies.length() > 1)
+	{
+		species = unique(genomes.individual.subpopulation.species, preserveOrder=F);
+		if (species.length() != 1)
+			stop("ERROR (calcTajimaD()): genomes must all belong to the same species.");
+		if (!isNULL(muts))
+			if (!all(muts.mutationType.species == species))
+				stop("ERROR (calcTajimaD()): muts must all belong to the same species as genomes.");
+	}
+	else
+	{
+		species = community.allSpecies;
+	}
+	
+	if (isNULL(muts))
+		muts = species.mutations;
+	
+	// handle windowing
+	if (!isNULL(start) & !isNULL(end))
+	{
+		if (start > end)
+			stop("ERROR (calcTajimaD()): start must be less than or equal to end.");
+		mpos = muts.position;
+		muts = muts[(mpos >= start) & (mpos <= end)];
+		length = end - start + 1;
+	}
+	else if (!isNULL(start) | !isNULL(end))
+	{
+		stop("ERROR (calcTajimaD()): start and end must both be NULL or both be non-NULL.");
+	}
+
+	// narrow down to the mutations that are actually present in the genomes and aren't fixed
+	p = genomes.mutationFrequenciesInGenomes(muts);
+	muts = muts[(p != 0.0) & (p != 1.0)];
+	
+	// do the calculation
+	// Pi and Watterson's theta functions divide by sequence length so this must be undone in Tajima's D
+	// Sequence length is constant (i.e. no missing data or indels) so this can be applied equally over both metrics
+	diff = (calcPiTheta(genomes) - calcWattersonsTheta(genomes)) * (species.chromosome.lastPosition + 1);
+	// calculate standard deviation of covariance of pi and Watterson's theta
+	k = size(muts);
+	n = genomes.size();
+	a_1 = sum(1 / 1:(n - 1));
+	a_2 = sum(1 / 1:(n - 1 ^ 2));
```

Comment:
```suggestion
	a_2 = sum(1 / 1:((n - 1) ^ 2));
```
... right?

Reply:
There is an issue here, thanks! But I now believe this should be the correct one and I should have it in my fork now:

a_2 = sum(1 / (1:(n - 1)) ^ 2);

I'm using Durrett 2008 section 2.4.1 for the exact formula. 1 / (n - 1) is a vector, all values in this vector are squared before the division from 1. I decided to test it more closely this time around with a population of only 10 individuals (20 genomes). The output should be equivalent to this python code:

1/19**2 + 1/18**2 + 1/17**2 + 1/16**2 + 1/15**2 + 1/14**2 + 1/13**2 + 1/12**2 + 1/11**2 +1/10**2 + 1/9**2 + +1.8**2 + 1/7**2 + 1/6**2 + 1/5**2 + 1/4**2 + 1/3**2 + 1/2**2 + 1/1**2

which equals 1.5780382439130234 and this is what I get with the above a_2. Both my original and your suggestion deviated from this. I guess that's the trickiest part to code with the vector stuff but I ran the calcTajimaD outputting every variable with this same small population to recheck and I'm more confident it's all working now. I guess I should upload a small script actually using these functions.

Reply:
Looks good; however,  note that I get the same value with your code and my suggestion (and indeed they should be equivalent).

Reply:
Ah, thanks for checking again. I'm trying now with two identical Tajima's D functions (except differing by your a_2 and mine) and only returning a_2, and indeed they are the same. I think a week ago I was trying to simplify it by running a function for both a_2's only and must've made some copy error. And I guess I got confused by the parentheses but I see how yours is basically just being more explicit that 1:(n - 1) is squared with the extra parentheses but the order of operations does work the same in both.